### PR TITLE
Implements Functional Options #1681

### PR DESCRIPTION
**Pull request by d2g**
You may wish to add additional defaults (i.e. UserAgent?).

### CODE REVIEW

**Comment by corylanou** [comment]
What is the reason we are adding this level of complexity?  Was there missing functionality somewhere?

**Comment by d2g** [comment]
Sorry I should have added the detail here rather than opening issue #1681 to try and explain it first.

Dave Cheney's post explains it better than I ever could:

http://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

File: client/influxdb.go [hunk 1]
```diff
@@ -27,6 +20,61 @@ type Client struct {
...
+// Set Password to be set as an option.
+func Password(p string) func(*Client) error {
+	return func(t *Client) error {
+		t.password = p
+		return nil
+	}
+}
+
+// Allow the Hostname to be configured.
+func URL(h url.URL) func(*Client) error {
```
Comment (corylanou) [inline client/influxdb.go]: nit: `(u url.URL)` instead of `(h url.URL)`

File: client/influxdb.go [hunk 2]
```diff
@@ -27,6 +20,61 @@ type Client struct {
...
+			Host:   "localhost:8086",
+		},
+		httpClient: &http.Client{},
+	}
+
+	return &c, c.setOption(options...)
+}
+
+// SetOption takes one or more option function and applies them in order to Client.
+func (t *Client) setOption(options ...func(*Client) error) error {
```
Comment (corylanou) [inline client/influxdb.go]: nit: `(c *Client)` instead of `(t *Client)`

**Comment by corylanou** [comment]
Ok, that makes perfect sense then.  I read that article when it came out and really liked it.  This is an excellent use of it, especially in the client facing api.

**Comment by otoolep** [comment]
Very interesting pattern @d2g - thanks for bringing it to our attention.

**Comment by toddboom** [comment]
I don't think we're planning to merge this, so I'm going to close it out. We might revisit this later as we revise the CLI.

**Pull request closed without merging**